File: 0663-equal-tree-partition/0663-equal-tree-partition.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def checkEqualTree(self, root: Optional[TreeNode]) -> bool:
        seen = []
        def dfs(node):
            if not node: return 0
            seen.append(node.val + dfs(node.left) + dfs(node.right))
            return seen[-1]
        total = dfs(root)
        seen.pop()
        for s in seen:
            if 2 * s == total: return True
        return False

File: 0663-equal-tree-partition/0663-equal-tree-partition.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Stack<Integer> seen;
    public boolean checkEqualTree(TreeNode root) {
        seen = new Stack<>();
        int total = dfs(root);
        seen.pop();
        for(int s : seen)
            if(s * 2 == total) return true;
        return false;
    }
    
    private int dfs(TreeNode root) {
        if(root == null) return 0;
        seen.push(root.val + dfs(root.left) + dfs(root.right));
        return seen.peek();
    }
}